### PR TITLE
dkg: add keymanager support

### PR DESCRIPTION
**Pull request by xenowits**
Adds support for keymanager API to `charon dkg` command.

category: feature
ticket: #1504

### CODE REVIEW

File: dkg/disk.go [hunk 1]
```diff
@@ -95,8 +98,43 @@ func loadDefinition(ctx context.Context, conf Config) (cluster.Definition, error
...
-func writeKeystores(datadir string, shares []share) error {
+// writeKeysToKeymanager writes validator private keyshares for the node to the provided keymanager address.
+func writeKeysToKeymanager(ctx context.Context, keymanagerURL string, shares []share) error {
+	var (
+		keystores []keystore.Keystore
+		passwords []string
+	)
+
+	for _, s := range shares {
+		password, err := testutil.RandomHex32()
```
Comment (corverroos) [inline dkg/disk.go:109]: can't use testutil in production code, also make sure you not using math.Random, but crypto.Random.

File: dkg/dkg.go [hunk 2]
```diff
@@ -73,6 +75,14 @@ func Run(ctx context.Context, conf Config) (err error) {
 		return err
 	}
 
+	// Check if keymanager address is reachable.
+	if conf.KeymanagerAddr != "" {
+		cl := keymanager.New(conf.KeymanagerAddr)
+		if err = cl.VerifyConnection(ctx); err != nil {
+			return err
```
Comment (corverroos) [inline dkg/dkg.go:82]: what does this error look like to users? please paste a screenshot

File: dkg/dkg.go [hunk 3]
```diff
@@ -191,10 +201,17 @@ func Run(ctx context.Context, conf Config) (err error) {
 	// Write keystores, deposit data and cluster lock files after exchange of partial signatures in order
 	// to prevent partial data writes in case of peer connection lost
 
-	if err := writeKeystores(conf.DataDir, shares); err != nil {
-		return err
+	if conf.KeymanagerAddr != "" { // Save to keymanager
+		if err = writeKeysToKeymanager(ctx, conf.KeymanagerAddr, shares); err != nil {
```
Comment (corverroos) [inline dkg/dkg.go:205]: what happens if keys have already been pushed in a previous DKG attempt?

File: dkg/dkg.go [hunk 4]
```diff
@@ -191,10 +201,17 @@ func Run(ctx context.Context, conf Config) (err error) {
...
+			return err
+		}
+		log.Debug(ctx, "Saved keyshares to keymanager")
```
Comment (corverroos) [inline dkg/dkg.go:208]: add URL ass param, suggest "import" as verb as that is what the keymannager API uses.

File: cmd/dkg.go [hunk 5]
```diff
@@ -55,6 +56,10 @@ this command at the same time.`,
 	return cmd
 }
 
+func bindKeymanagerAddrFlag(flags *pflag.FlagSet, addr *string) {
+	flags.StringVar(addr, "keymanager-address", "", "The keymanager URL to push validator keyshares to.")
```
Comment (corverroos) [inline cmd/dkg.go:60]: suggest "import" as verb as that what keymanager uses.

File: dkg/dkg_test.go [hunk 6]
```diff
@@ -405,3 +408,134 @@ func startNewDKG(t *testing.T, parentCtx context.Context, config dkg.Config, dkg
...
+	require.NoError(t, def.VerifySignatures())
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	// Setup
+	dir, err := os.MkdirTemp("", "")
+	require.NoError(t, err)
+
+	allReceived := make(chan struct{}, 1) // Receives struct{} if all `numNodes` keystores are intercepted by the keymanager server
```
Comment (corverroos) [inline dkg/dkg_test.go:427]: why buffer of 1? unbuffered is default

File: dkg/dkg_test.go [hunk 7]
```diff
@@ -405,3 +408,134 @@ func startNewDKG(t *testing.T, parentCtx context.Context, config dkg.Config, dkg
 
 	return cancel
 }
+
+// TestKeymanager tests the keymanager functionality where the dkg command pushes validator keyshares
+// to a keymanager API url.
+func TestKeymanager(t *testing.T) {
```
Comment (corverroos) [inline dkg/dkg_test.go:414]: this feels like a duplicate test, can't we add this to existing table test?

File: dkg/dkg_test.go [hunk 8]
```diff
@@ -405,3 +408,134 @@ func startNewDKG(t *testing.T, parentCtx context.Context, config dkg.Config, dkg
...
+	Passwords []string       `json:"passwords"`
+}
+
+// noopKeystore is a mock keystore.keystore for use in tests.
+type noopKeystore struct {
+	Crypto map[string]interface{} `json:"crypto"`
+}
+
+// decrypt returns the secret from the encrypted keystore.
+func decrypt(t *testing.T, store noopKeystore, password string) (*bls_sig.SecretKey, error) {
```
Comment (corverroos) [inline dkg/dkg_test.go:505]: this feels like testing the keymanager client, which isn't required, you test that implementation in the keymanager package, this test is for DKG, testing that is submits something is fine.

**Review by corverroos — APPROVED** [review]

File: dkg/disk.go (hunk 1 above)
Comment (xenowits) [reply dkg/disk.go:109]: gotcha. The random function's using crypto/rand though it shouldn't be part of the testutil package. Will refactor this out into `eth2util` in a separate PR

File: dkg/dkg.go (hunk 2 above)
Comment (xenowits) [reply dkg/dkg.go:82]: Screenshot 2023-01-24 at 15 29 16

Should we have `cannot ping keymanager address` in error log?

File: dkg/dkg.go (hunk 3 above)
Comment (xenowits) [reply dkg/dkg.go:205]: A new DKG attempt creates fresh keys so new keys are pushed to keymanager and previously pushed keys would not be overwritten (are not affected).

File: dkg/dkg.go (hunk 4 above)
Comment (xenowits) [reply dkg/dkg.go:208]: Screenshot 2023-01-24 at 16 00 32

File: dkg/dkg_test.go (hunk 7 above)
Comment (xenowits) [reply dkg/dkg_test.go:414]: yeah we can, doing this now

File: dkg/dkg_test.go (hunk 8 above)
Comment (xenowits) [reply dkg/dkg_test.go:505]: cool, removing this. I'm now only ensuring that the request body received by the server is non-empty

File: dkg/dkg_test.go (hunk 6 above)
Comment (xenowits) [reply dkg/dkg_test.go:427]: i initially intended to make it unblocking, but i have updated the tests and made this unbuffered

File: dkg/dkg.go (hunk 2 above)
Comment (corverroos) [reply dkg/dkg.go:82]: suggest: `return errors.Wrap(err, "verify keymanager address"`

File: dkg/dkg_test.go [hunk 9]
```diff
@@ -405,3 +440,35 @@ func startNewDKG(t *testing.T, parentCtx context.Context, config dkg.Config, dkg
...
+		mu.Unlock()
+
+		t.Logf("Received keystore: %d/%d\n", cnt, numNodes)
+		if cnt == numNodes {
+			go func() {
+				allReceived <- struct{}{}
+			}()
+		}
+
+		w.WriteHeader(http.StatusOK)
```
Comment (corverroos) [inline dkg/dkg_test.go:472]: no need for this, can just remove

File: dkg/dkg_test.go [hunk 10]
```diff
@@ -90,6 +113,14 @@ func testDKG(t *testing.T, def cluster.Definition, p2pKeys []*ecdsa.PrivateKey)
 		TestDef: &def,
 	}
 
+	allReceived := make(chan struct{}) // Receives struct{} if all `numNodes` keystores are intercepted by the keymanager server
+	if keymanager {
+		srv := httptest.NewServer(newKeymanagerHandler(t, len(def.Operators), allReceived))
```
Comment (corverroos) [inline dkg/dkg_test.go:118]: I would just inline a very simple handler:
```
httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
    receives <- struct{}
  }
}
```
and then do below:
```
for expectReceives > 0 {
   <-receives
   expectReceives--
}
```
```

File: dkg/disk.go [hunk 11]
```diff
@@ -199,3 +237,14 @@ func validURI(str string) bool {
 
 	return err == nil && (u.Scheme == "http" || u.Scheme == "https") && u.Host != ""
 }
+
+// randomHex32 returns a random 32 character hex string. It uses crypto/rand.
+func randomHex32() (string, error) {
```
Comment (corverroos) [inline dkg/disk.go:242]: need to do this for createcluster as well

**Review by corverroos — APPROVED** [review]

**Pull request merged**